ops/testing.py

```diff
@@ -446,16 +444,19 @@ def _next_relation_id(self):
         self._relation_id_counter += 1
         return rel_id
 
-    def add_storage(self, storage_name: str, count: int = 1) -> typing.List[str]:
-        """Declare a new storage device attached to this unit.
+    def add_storage(self, storage_name: str, count: int = 1,
+                    attach: bool = False) -> typing.List[str]:
+        """Create a new storage device and attach it to this unit.
 
         To have repeatable tests, each device will be initialized with
-        location set to /<storage_name>N, where N is the counter and
-        will be a number from [0,total_num_disks-1]
+        location set to /[tmpdir]/<storage_name>N, where N is the counter and
+        will be a number from [0,total_num_disks-1].
 
         Args:
             storage_name: The storage backend name on the Charm
             count: Number of disks being added
+            attach: True to also attach the storage mount and emit storage-attached if
+                harness.begin() has been called.
```

Comment:
> if harness.begin() has been called.

Seems like it should raise if `attach=True` before `begin`, i.e.

```suggestion
            attach: Flag indicating whether to also attach the storage mount.
        
        Raises:
            RuntimeError, if called before `begin`.
```

Reply:
I could go both ways.  Allowing attach before begin does give more control to a charm dev to skip storage-attached events when they do call begin.  But I suppose they could just call disable_hooks to achieve a similar thing.  I'll think about this one a bit more.